File: app/src/main/java/com/moyinoluwa/facedetection/MainActivity.java
package com.moyinoluwa.facedetection;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.SparseArray;
import android.view.View;
import android.widget.ImageView;

import com.google.android.gms.vision.Frame;
import com.google.android.gms.vision.face.Face;
import com.google.android.gms.vision.face.FaceDetector;
import com.google.android.gms.vision.face.Landmark;

import java.util.concurrent.Callable;

import rx.Observable;
import rx.Subscriber;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity {
    private ImageView imageView;
    private Paint rectPaint;
    private Bitmap defaultBitmap;
    private Bitmap temporaryBitmap;
    private Bitmap eyePatchBitmap;
    private Canvas canvas;
    private Subscription faceDetectorSubscription;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        imageView = (ImageView) findViewById(R.id.image_view);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        if (faceDetectorSubscription != null) {
            faceDetectorSubscription.unsubscribe();
        }
    }

    public void processImage(View view) {

        BitmapFactory.Options bitmapOptions = new BitmapFactory.Options();
        bitmapOptions.inMutable = true;

        initializeBitmap(bitmapOptions);
        createRectanglePaint();

        canvas = new Canvas(temporaryBitmap);
        canvas.drawBitmap(defaultBitmap, 0, 0, null);

        final FaceDetector faceDetector = new FaceDetector.Builder(this)
                .setTrackingEnabled(false)
                .setLandmarkType(FaceDetector.ALL_LANDMARKS)
                .build();

        if (!faceDetector.isOperational()) {
            new AlertDialog.Builder(this)
                    .setMessage("Face Detector could not be set up on your device :(")
                    .show();
        } else {
            Frame frame = new Frame.Builder().setBitmap(defaultBitmap).build();

            faceDetectorSubscription = getSubscription(faceDetector, frame);
        }
    }

    private void initializeBitmap(BitmapFactory.Options bitmapOptions) {
        defaultBitmap = BitmapFactory.decodeResource(getResources(), R.drawable.image,
                bitmapOptions);
        temporaryBitmap = Bitmap.createBitmap(defaultBitmap.getWidth(), defaultBitmap
                .getHeight(), Bitmap.Config.RGB_565);
        eyePatchBitmap = BitmapFactory.decodeResource(getResources(), R.drawable.eye_patch,
                bitmapOptions);
    }

    private void createRectanglePaint() {
        rectPaint = new Paint();
        rectPaint.setStrokeWidth(5);
        rectPaint.setColor(Color.CYAN);
        rectPaint.setStyle(Paint.Style.STROKE);
    }

    private void detectFaces(SparseArray<Face> sparseArray) {

        for (int i = 0; i < sparseArray.size(); i++) {
            Face face = sparseArray.valueAt(i);

            float left = face.getPosition().x;
            float top = face.getPosition().y;
            float right = left + face.getWidth();
            float bottom = right + face.getHeight();
            float cornerRadius = 2.0f;

            RectF rectF = new RectF(left, top, right, bottom);
            canvas.drawRoundRect(rectF, cornerRadius, cornerRadius, rectPaint);

            detectLandmarks(face);
        }
    }

    private void detectLandmarks(Face face) {
        for (Landmark landmark : face.getLandmarks()) {

            int cx = (int) (landmark.getPosition().x);
            int cy = (int) (landmark.getPosition().y);

            canvas.drawCircle(cx, cy, 10, rectPaint);

            drawLandmarkType(landmark.getType(), cx, cy);

            drawEyePatchBitmap(landmark.getType(), cx, cy);
        }
    }

    private void drawLandmarkType(int landmarkType, float cx, float cy) {
        String type = String.valueOf(landmarkType);
        rectPaint.setTextSize(50);
        canvas.drawText(type, cx, cy, rectPaint);
    }

    private void drawEyePatchBitmap(int landmarkType, float cx, float cy) {

        if (landmarkType == 4) {
            // TODO: Optimize so that this calculation is not done for every face
            int scaledWidth = eyePatchBitmap.getScaledWidth(canvas);
            int scaledHeight = eyePatchBitmap.getScaledHeight(canvas);
            canvas.drawBitmap(eyePatchBitmap, cx - (scaledWidth / 2), cy - (scaledHeight / 2), null);
        }
    }

    private Observable<SparseArray> getFaceDetectorObservable(final FaceDetector faceDetector,
                                                              final Frame frame) {
        return Observable.fromCallable(new Callable<SparseArray>() {

            @Override
            public SparseArray call() throws Exception {
                SparseArray<Face> sparseArray = faceDetector.detect(frame);
                detectFaces(sparseArray);
                return null;
            }
        });
    }

    private Subscription getSubscription(final FaceDetector faceDetector, Frame frame) {
        return getFaceDetectorObservable(faceDetector, frame)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<SparseArray>() {
                    @Override
                    public void onCompleted() {
                        imageView.setImageDrawable(new BitmapDrawable(getResources(),
                                temporaryBitmap));
                        faceDetector.release();
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onNext(SparseArray sparseArray) {

                    }
                });
    }
}

File: app/src/main/java/com/moyinoluwa/facedetection/MainActivity.kt
package com.moyinoluwa.facedetection

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.drawable.BitmapDrawable
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.util.SparseArray
import android.view.View
import android.widget.ImageView
import com.google.android.gms.vision.Frame
import com.google.android.gms.vision.face.Face
import com.google.android.gms.vision.face.FaceDetector

private const val LEFT_EYE = 4
private const val RADIUS = 10f
private const val TEXT_SIZE = 50f
private const val CORNER_RADIUS = 2f
private const val STROKE_WIDTH = 5f
class MainActivity : AppCompatActivity() {

    lateinit var imageView: ImageView
    lateinit var defaultBitmap: Bitmap
    lateinit var temporaryBitmap: Bitmap
    lateinit var eyePatchBitmap: Bitmap
    lateinit var canvas: Canvas

    val rectPaint = Paint()
    val faceDetector: FaceDetector
        get() = initializeDetector()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        imageView = findViewById<View>(R.id.image_view) as ImageView
    }

    fun processImage(view: View) {
        val bitmapOptions = BitmapFactory.Options().apply {
            inMutable = true
        }

        initializeBitmap(bitmapOptions)
        createRectanglePaint()

        canvas = Canvas(temporaryBitmap).apply {
            drawBitmap(defaultBitmap, 0f, 0f, null)
        }

        if (!faceDetector.isOperational) {
            AlertDialog.Builder(this)
                       .setMessage("Face Detector could not be set up on your device :(")
                       .show()
        } else {
            val frame = Frame.Builder().setBitmap(defaultBitmap).build()
            val sparseArray = faceDetector.detect(frame)

            detectFaces(sparseArray)

            imageView.setImageDrawable(BitmapDrawable(resources, temporaryBitmap))

            faceDetector.release()
        }
    }

    private fun initializeBitmap(bitmapOptions: BitmapFactory.Options) {
        defaultBitmap = BitmapFactory.decodeResource(resources, R.drawable.image, bitmapOptions)
        temporaryBitmap = Bitmap.createBitmap(defaultBitmap.width,
                                              defaultBitmap.height,
                                              Bitmap.Config.RGB_565)
        eyePatchBitmap = BitmapFactory.decodeResource(resources, R.drawable.eye_patch, bitmapOptions)
    }

    private fun createRectanglePaint() {
        rectPaint.apply {
            strokeWidth = STROKE_WIDTH
            color = Color.CYAN
            style = Paint.Style.STROKE
        }
    }

    private fun initializeDetector(): FaceDetector {
        return FaceDetector.Builder(this)
                           .setTrackingEnabled(false)
                           .setLandmarkType(FaceDetector.ALL_LANDMARKS)
                           .build()
    }

    private fun detectFaces(sparseArray: SparseArray<Face>) {
        for (i in 0 until sparseArray.size()) {
            val face = sparseArray.valueAt(i)

            val left = face.position.x
            val top = face.position.y
            val right = left + face.width
            val bottom = top + face.height

            val rectF = RectF(left, top, right, bottom)
            canvas.drawRoundRect(rectF, CORNER_RADIUS, CORNER_RADIUS, rectPaint)

            detectLandmarks(face)
        }
    }

    private fun detectLandmarks(face: Face) {
        for (landmark in face.landmarks) {
            val xCoordinate = landmark.position.x
            val yCoordinate = landmark.position.y

            canvas.drawCircle(xCoordinate, yCoordinate, RADIUS, rectPaint)

            drawLandmarkType(landmark.type, xCoordinate, yCoordinate)
            drawEyePatchBitmap(landmark.type, xCoordinate, yCoordinate)
        }
    }

    private fun drawLandmarkType(landmarkType: Int, xCoordinate: Float, yCoordinate: Float) {
        val type = landmarkType.toString()
        rectPaint.textSize = TEXT_SIZE
        canvas.drawText(type, xCoordinate, yCoordinate, rectPaint)
    }

    private fun drawEyePatchBitmap(landmarkType: Int, xCoordinate: Float, yCoordinate: Float) {
        when (landmarkType) {
            LEFT_EYE -> {
                // TODO: Optimize so that this calculation is not done for every face
                val scaledWidth = eyePatchBitmap.getScaledWidth(canvas)
                val scaledHeight = eyePatchBitmap.getScaledHeight(canvas)
                canvas.drawBitmap(eyePatchBitmap,
                                  xCoordinate - scaledWidth / 2,
                                  yCoordinate - scaledHeight / 2,
                                  null)
            }
        }
    }
}